legate/core/context.py

```diff
@@ -66,6 +66,35 @@ def find_last_user_frame(libname: str) -> str:
     return f"{frame.f_code.co_filename}:{frame.f_lineno}"
 
 
+class LibraryAnnotations:
+    def __init__(self) -> None:
+        self._entries: dict[str, str] = {}
+
+    @property
+    def empty(self) -> bool:
+        return len(self._entries) == 0
+
+    def __getitem__(self, key: str) -> Optional[str]:
+        return self._entries[key] if key in self._entries else None
+
+    def __setitem__(self, key: str, value: str) -> None:
+        self._entries[key] = value
+
+    def remove(self, key: str) -> None:
+        del self._entries[key]
+
+    def __str__(self) -> str:
+        return "|".join(
+            f"{key},{value}" for key, value in self._entries.items()
+        )
+
+    def __repr__(self) -> str:
+        return str(self)
+
+
+PROVENANCE_KEY = "Provenance"
```

Comment:
I usually tack on some dunders for good measure if I am reserving a "special" key name that I don't want to collide with anything a user might supply

Reply:
Though as long as there is already a class `LibraryAnnotations` couldn't the provenance be stored separately instead of in the same dict?

Reply:
https://github.com/nv-legate/legate.core/pull/464/commits/828d27955457b8038419ef37e13dac281118a7ac

Reply:
took the suggestion that the provenance can be separate